### ISSUE TITLE
Review notes

### ISSUE THREAD

**Issue by JiriLojda** [Enngage/kontent-data-manager#1]
Review notes:
* https://github.com/Enngage/kontent-data-manager/blob/main/README.md#how-it-works 
  * It shouldn't be a problem to check the two specified projects have to same models and stuff. Since it is essential, it would perhaps be a good idea to do that before starting the import/export itself.
* https://github.com/Enngage/kontent-data-manager/blob/main/tslint.json
  * Tslint is deprecated for a long time. Use eslint instead.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/core/translation-helper.ts#L374
  * It would be good to have the transformation (from delivery format to MAPI format) separated into a reusable package. We can even use the reusable richtext parser once it is done https://github.com/pokornyd/kontent-rich-text-to-json-converter.
  * Also, each of the comment-separated parts should be a separate function. It is, imho, more readable and reusable then comment-separated pieces of code. https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/core/translation-helper.ts#L455-L458
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/core/translation-helper.ts#L35
  * Why does the core helper reference something related to the output format? At least I presume the `csv` in the name means the file format. I expect this code to be independent on the output format as I can use a different format then csv.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/core/translation-helper.ts#L38
  * If you'd use an object instead of an array with type like so `Readonly<Record<ElementType, TransformFnc>>`, you would not only get a better performance when searching for the matching transformer. Most importantly, TypeScript would make sure, that each element type has exactly one transformer. Then, if we can either relay on Kontent.ai always returning a correct `ElementType` and there would be no need to check a transformer exists because that would be ensured at compile-time. Or the check can remain, but it would only be triggered in an unlikely case of Kontent.ai returning an invalid `ElementType`. Either way less opportunity for a mistake in the code here.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/core/translation-helper.ts#L492
  * Why is it a class? It's just a collections of exported or non-exported functions and some internal constants.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/core/translation-helper.ts#L258-L268
  * Removing the curly braces in those simple functions would make them much more readable in my opinion.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/export/export.service.ts#L109-L120
  * The custom export function has no option to load components? (Marked as `itemType` `'component'`)
  * Update: Apparently, the difference is only in a slightly different logging output into the console so not a big deal then.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/export/export.service.ts#L131
  * Wouldn't it be more efficient to use the `.types` method and load all types at once (I know it could be multiple requests but at least a single stream of continuations) instead of iterating it? I am not sure if something similar exists for languages.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/export/export.service.ts#L37-L38
  * These ternary operators can be simplified to `!!yourVariable`.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/export/export.service.ts#L183-L185
  * The name of the functions is rather misleading, in my opinion. It only logs something and doesn't even use all of its arguments. Do you have more plans for the functions?
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/import/import.service.ts#L105-L136
  * You can replace each of the for-cycles with a single `.filter` call that will be calling the `config.canImport` functions. The number of removed items/assets can then be easily computed by subtracting the filtered array length from the original array length. As a bonus to being more readable, in my opinion, the algorithm will be linear instead of quadratic.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/import/import.service.ts#L145-L181
  * This should be a separate function in my opinion.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/import/import.service.ts#L732
  * It would be better to write error messages to the error output with `console.error`.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/import/import.service.ts#L736-L762
  * I believe it would be easier to understand if this method would be separated into a function the creates the import data object and a function that logs info each with a relevant name.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/import/import.service.ts#L764
  * This doesn't seem to be used.
 * https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/import/import.service.ts#L223-L236
   * I am not sure I understand this code. You push the same data into the array twice. This will be much simpler if you split the method as I suggested above.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/import/import.service.ts#L542-L544
  * If you'd use `.then(r => r.data)` instead, there would be no need for wrapping the awaited statement with braces and the formatting would look a bit better imho. 
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/import/import.service.ts#L557-L573
  * Combining the `if`s and negating them would help you avoid this Egyptian pyramid.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/import/import.service.ts#L269
  * It would be better to have the item-updating and variant-updating part separated. The method is already very long.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/import/import.models.ts#L81
  * Is this necessary?
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/import/import.service.ts#L363-L366
  * If you need to add comments to parts of a function to make it clear what it's doing, it most likely should be a separate function. This moving to a workflow step is such a case imho.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/import/import.service.ts#L449-L470
  * The error-handling logic could also be separated and reused.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/import/import.service.ts#L479-L535
  * The two methods are almost the same. I would consider unifying them into one that accepts a predicate as an argument. The predicate would decide whether the current workflow is the one you are looking for or not.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/file-processor/file-processor.service.ts#L179-L185
  * This can be as simple as `const zipSizeInBytes: number = zipData instanceof Blob ? zipData.size : zipData.byteLength;`.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/file-processor/file-processor.service.ts#L204-L206
  * To be precise this function returns type `Promise<number>`, but semantically I would say it returns type `Promise<void>`. I would certainly avoid using type `any`. Same to the parameter of the callback you pass into the `Promise` constructor. (i.e. `resolve`) You don't need to supply it with an explicit type. TypeScript is capable of inferring it from the type of the `Promise` constructor. If you'd really like to provide it with an explicit type a more appropriate type is `(res: T) => void` where `T` is the type in the promise, in this case `number`, but I think it is much better to let the TypeScript infer the type here.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/file-processor/file-processor.service.ts#L328-L335
  * `isFileFromDirectory(dirName: string)` function comes to mind.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/file-processor/file-processor.service.ts#L327
  * Why do you always iterate over the object's entries when you only want values?
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/node/cli/app.ts#L86
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/node/cli/app.ts#L120-L131
  * I am not sure I understand this, but it seems to me, that you can only export a `.zip` file (the csv or json file would be within the zip), but can import `.zip` or `.csv` or `.json` as the root file. Am I correct? I guess there is an option missing to export just a single `.json` or `.csv` file.
  * I am also not very sure about the role of the `file-processor.service`. To me it seems it is a zip-file-handler with some csv and json capabilities inserted in. What would you think about having a file-type handler files (zip, csv, json...) each being able to export/import to/from a relevant file type. The zip handler would need to have a parameter that would create/parse files that are inside, but it would not assume it is a csv or json. The handler would be directly called in the app based on the file extension of the imported file or the option for export.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/file-processor/file-processor.models.ts#L44
  * This doesn't seem to be very safe. What if some element has `name` or `type` as codename?
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/file-processor/formats/base-processor.service.ts#L26
  * I don't find the name of this method very descriptive. It not only transforms the items into a different model (I guess a flattened model, I've written my concerns about the mixing of elements with other properties above). It also transforms element values using the translation helper. I find this rather surprising considering the name. I wonder whether the translation using the translation helper is something we want to do for all formats, as otherwise we won't be able to, for example, import the richtext through MAPI as it has different format after export from Deliver. If we would like to always use the translation, I would argue it would be better to call it somewhere above so that different file formats wouldn't need to worry about this implementation detail.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/file-processor/formats/csv-processor.service.ts#L73
  * Both parts of the `&&` condition seem to be the same. I suppose that is a mistake.
  * Also, it would be nice to collocate the definition of the special element header (creation, parsing and checking).
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/file-processor/formats/csv-processor.service.ts#L53-L56
  * Would it be possible to extract the header row outside the `foreach` cycle and scrap the index tracking? Something like `const headerRow = parser.read()`;
  * Another thing is whether we want to validate the header row in any way. First of there needs to be the row (no null or anything). Secondly, you expect there is the set of defined columns (i.e. `type`, `name`...). Should we validate they are present and inform the user that the file is invalid because of ...?
  * Another thing is how would we approach breaking changes of the output format in the future. For example we could create a backup transformer for each breaking change that updates an existing backup into the newer version of the backup manager.
  * It would also be nice to have the creation of an item from a single row (plus header row) in a separate function.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/file-processor/formats/csv-processor.service.ts#L209-L218
  * The desctructuring into a new array is redundant. The `.map` function already creates a new array on it's own.
  * I don't understand the code. Why is the value of the field the same as its name?
  * It would be better to use `unknown` instead of `any`. That applies to almost all usages of `any`. The best option is to use a specific type you really return so that TypeScript can really do its job. If you don't know the type you are returning, `unknown` is better than `any`, because with `any` you are basically turning TypeScript of for that property/variable/constant. With `unknown`, on the other hand, you are telling TypeScript that you don't know what the type is and it will make sure, that you will assert the type at runtime before you use the reference anywhere. Technically, `unknown` is not assignable to anything and if you want to use it in any way you need to assert it to a certain type ideally with type guards.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/file-processor/formats/csv-processor.service.ts#L142-L144
  * If the type of the columnNames would be `('field1' | 'field2')[]` or even better `['field1', 'field2']`, you wouldn't need to convert the `parsedAsset` to `any` and therefore use the type safety. Of course that would mean to validate the field names properly which is a good idea anyway. 
  * It would also be good idea (similarly to content items) to extract the conversion of a line into an asset into a separate function.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/file-processor/formats/csv-processor.service.ts#L191-L196
  * Or you can write `filter(m => !!m.codename?.length)`.
* The `file-helper.ts` doesn't seem to be used.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/node/file/file.service.ts#L6
  * The `config` parameter doesn't seem to be used.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/core/global-helper.ts#L17-L20
  * You can inline this and simplify with the help of mr. De Morgan.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/core/global-helper.ts#L40
  * This doesn't seem to be used.
* https://github.com/Enngage/kontent-data-manager/blob/5161039878302a54bd86a030f6e9bd56fc57c9fc/lib/core/global-helper.ts#L52-L63
  * You can simplify this by throwing inside the `if` like so. `if (condition) {throw {...}; } throw error;`
* https://github.com/Enngage/kontent-data-manager/blob/main/samples/sample-custom-processor.service.ts#L33
  * You can specify a better type than this.
* I also think it might not be terribly complicated to add an option that would copy the same content type structure from the source project to the target project. I did it for a couple of content types manually and it is very long and boring work. Or is there a better way? I am not sure if project cloning could help here, but I can even clone items with that.
* I am having a trouble running the manager on my project. It fails to import an item with a content component with an error like this "The value of the property 'id' must be a valid GUID identifier.". It seems the component id is malformed with an extra "n" in the beginning. (n15829974-9f34-0176-b053-d17111bb9bab) Not sure if I am doing something wrong. The problem seems to occur for both csv and json backups.
* https://github.com/Enngage/kontent-data-manager/blob/main/README.md?plain=1#L64
  * This seems to be wrong. When I try to run it, the program complains that "Unsupported export format 'undefined'".
* It is confusing that when I export data in the json format, the name of the final zip file is named `csv-...`. Also, I think I mentioned it earlier, I expected to be able to export into a pure json, not a zip with json files.

**Comment by Enngage** [comment]
@JiriLojda, thank you very much for such a thorough review :) I'll be going through that as soon as I can.

**Comment by Enngage** [comment]
It took some time, but I've finally been able to go back to this :) Thanks again @JiriLojda for very valuable input. I've implemented most of the recommendations :) 

There are some things that this library doesn't intend to do such as using Management API to sync models between projects (which is actually really difficult to do via API). However, It should solve scenarios such as "export data fast in a certain format" and "import content from a certain format".
